### PR TITLE
Stapler 1.253

### PR DESCRIPTION
**Pull request by jglick**
[Changelog](https://github.com/stapler/stapler/blob/master/CHANGELOG.md#1253) and [diffs](https://github.com/stapler/stapler/compare/stapler-parent-1.252...stapler-parent-1.253).

Also [Commons Codec changelog](https://commons.apache.org/proper/commons-codec/changes-report.html#a1.9).

### Proposed changelog entries

* Stapler library upgraded to 1.253:
  * Improved performance for some Blue Ocean operations.
  * Various changes which could be of interest to plugin developers, for which see component changelog.
* Commons Codec library upgraded to 1.9.

### Desired reviewers

@jenkinsci/code-reviewers @reviewbybees

### CODE REVIEW

**Review by oleg-nenashev — APPROVED** [review]
🐝 for the integration, but the changelog entry could be improved a bit. E.g. the binary compatibility may worth mentioning

**Review by oleg-nenashev — DISMISSED** [review]
There are Enforcer issues:

```
[WARNING] Rule 3: org.apache.maven.plugins.enforcer.RequireUpperBoundDeps failed with message:

Failed while enforcing RequireUpperBoundDeps. The error(s) are [

Require upper bound dependencies error for commons-codec:commons-codec:1.8 paths to dependency are:

+-org.jenkins-ci.main:jenkins-core:2.86-SNAPSHOT
  +-commons-codec:commons-codec:1.8
and
+-org.jenkins-ci.main:jenkins-core:2.86-SNAPSHOT
  +-org.jenkins-ci.main:cli:2.86-SNAPSHOT
    +-commons-codec:commons-codec:1.4
and
+-org.jenkins-ci.main:jenkins-core:2.86-SNAPSHOT
  +-commons-httpclient:commons-httpclient:3.1-jenkins-1
    +-commons-codec:commons-codec:1.8
and
+-org.jenkins-ci.main:jenkins-core:2.86-SNAPSHOT
  +-org.acegisecurity:acegi-security:1.0.7
    +-commons-codec:commons-codec:1.3
and
+-org.jenkins-ci.main:jenkins-core:2.86-SNAPSHOT
  +-org.kohsuke.stapler:stapler-jrebel:1.253
    +-org.kohsuke.stapler:stapler:1.253
      +-commons-codec:commons-codec:1.9
]

```

**Comment by oleg-nenashev** [comment]
Commons Codec 1.9 has been introduced by me in https://github.com/stapler/stapler/commit/4d59a74413f40f7064c8486d508684d38ebdfc38, upper bounds in Stapler

**Comment by oleg-nenashev** [comment]
I did it in the Stapler Core because of the following:

```
Require upper bound dependencies error for commons-codec:commons-codec:1.8 paths to dependency are:
+-org.kohsuke.stapler:stapler:1.254-SNAPSHOT
  +-commons-codec:commons-codec:1.8
and
+-org.kohsuke.stapler:stapler:1.254-SNAPSHOT
  +-org.jvnet.hudson:htmlunit:2.6-hudson-2
    +-commons-codec:commons-codec:1.4
and
+-org.kohsuke.stapler:stapler:1.254-SNAPSHOT
  +-org.jvnet.hudson:htmlunit:2.6-hudson-2
    +-commons-httpclient:commons-httpclient:3.1
      +-commons-codec:commons-codec:1.2
and
+-org.kohsuke.stapler:stapler:1.254-SNAPSHOT
  +-org.apache.httpcomponents:httpmime:4.5
    +-org.apache.httpcomponents:httpclient:4.5
      +-commons-codec:commons-codec:1.9
```

File: core/src/main/java/jenkins/slaves/EncryptedSlaveAgentJnlpFile.java [hunk 1]
```diff
@@ -70,7 +70,7 @@ public void generateResponse(StaplerRequest req, StaplerResponse res, Object nod
             final ByteArrayOutputStream baos = new ByteArrayOutputStream();
             StaplerResponse temp = new ResponseImpl(req.getStapler(), new HttpServletResponseWrapper(res) {
                 @Override public ServletOutputStream getOutputStream() throws IOException {
-                    return new FilterServletOutputStream(baos);
+                    return new FilterServletOutputStream(baos, super.getOutputStream());
```
Comment (oleg-nenashev) [inline core/src/main/java/jenkins/slaves/EncryptedSlaveAgentJnlpFile.java]: Could it cause issues for downstream deps?

Comment (jglick) [reply core/src/main/java/jenkins/slaves/EncryptedSlaveAgentJnlpFile.java]: I guess? Ask @jtnord who requested this.

Comment (jglick) [reply core/src/main/java/jenkins/slaves/EncryptedSlaveAgentJnlpFile.java]: See discussion: https://github.com/stapler/stapler/pull/131#discussion_r140082051

Comment (jtnord) [reply core/src/main/java/jenkins/slaves/EncryptedSlaveAgentJnlpFile.java]: this is the wrong change IIUC.  that wraps the  output stream of the response - which will return false information.

The correct fix is https://github.com/jenkinsci/jenkins/pull/3033/files#diff-8991f368b3f7346bf83c7081c719c809R118

**Comment by jtnord** [comment]
I beleive this is missing other changes for dependencies to align with stapler and servlet 3.x spec 
see https://github.com/jenkinsci/jenkins/pull/3033

File: core/src/main/java/jenkins/slaves/EncryptedSlaveAgentJnlpFile.java (hunk 1 above)
Comment (jtnord) [reply core/src/main/java/jenkins/slaves/EncryptedSlaveAgentJnlpFile.java]: FilterServletOutputStream is likely to be removed in a future version - its only there to do home grown GZIP which is causing more harm that good, and other places have abused it to do things like this...

**Comment by oleg-nenashev** [comment]
@jtnord @jglick Would it be possible to prevent the binary compatibility breakage by adding a bridge method somehow?

**Comment by jglick** [comment]
> adding a bridge method

-1, too crazy. We could restore and deprecate the original constructor (and just have the new methods throw `UnsupportedOperationException` or something). But it looks like this PR just needs to be replaced by #3033 (after that bumps up to the release version), right @jtnord?

**Comment by jtnord** [comment]
@jglick I think we should close this in favour of #3033

**Comment by jglick** [comment]
OK.

**Pull request closed without merging**